### PR TITLE
Fix ray spells: ripple effect for disrupting ray and correct drawing

### LINKED ISSUE #645: "Cold Ray" spell animation issue
I've noticed, that we have some problems, while animating "Cold Ray" spell:

The closer unit we attack using this spell, the more squeezed animation we have:
Cold 2 long
Cold 2
Up to overlapping each circle:
Cold short

Can our esteemed team look at this issue?

### PR DESCRIPTION
**Pull request by idshibanov**
Fixes #645. 

Ripple function is very magical. We should reuse it later: teleport effect uses similar logic.

### CODE REVIEW

**Comment by idshibanov** [comment]
ezgif-5-b0d5c39735fe

File: src/engine/surface.cpp [hunk 1]
```diff
@@ -1113,6 +1114,35 @@ Surface Surface::RenderSepia( void ) const
...
+
+    const int linearWave = abs( 20 - ( frame + 10 ) % 40 ) - 10;
+    const int progress = 7 - ( frame / 10 );
+    const double xOffset = ( progress * 0.05 + 0.3 ) * linearWave;
+
+    for ( int y = 0; y < height; ++y ) {
+        double sinYEffect = ( sin( ( y % 62 ) / 20.0 ) - 0.5 ) * 2.0;
+        for ( int x = 0; x < width; ++x ) {
+            const int newX = x + static_cast<int>( xOffset * sinYEffect );
+            if ( newX >= 0 && newX < width ) {
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/engine/surface.cpp:1135]: I was thinking about that but not too worried. Most ICN sprites have a lot of room in them. Part of the charm of the effect is distortion is gives, so if we truncate a bit it's ok.

If you want I can resize the surface, sinYEffect is in range of -1 to 1.

File: src/engine/surface.cpp [hunk 2]
```diff
@@ -1113,6 +1114,35 @@ Surface Surface::RenderSepia( void ) const
...
+Surface Surface::RenderRippleEffect( int frame ) const
+{
+    Surface res = GetSurface();
+
+    res.Lock();
+
+    const int height = h();
+    const int width = w();
+
+    const int linearWave = abs( 20 - ( frame + 10 ) % 40 ) - 10;
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/engine/surface.cpp:1127]: Agreed, I'll try to simplify it again. Two or three of these might be useful as a parameter.

**Comment by idshibanov** [comment]
@ihhub Added two control variables: `scaleX` decides how far it stretches on X axis and `waveFrequency` controls how often it will oscillate on Y axis.

I'm pre-scaling the surface, so there's no truncation anymore.

**Comment by ihhub** [comment]
Hi @idshibanov , thank you very much for your non-stop contribution!

**Pull request merged**